### ISSUE TITLE
Mouse.SetPosition not updating Mouse.GetState().Position on Mac

### ISSUE THREAD

**Issue by Torshid** [mono/MonoGame#4759]
I searched for a while regarding this issue and it seems that is has not been reported yet.
The Mouse.GetState() is only updated when the user moves the mouse himself.

Therefore, the output for this is the last position the user moved the mouse to, and not (0,0):
`Mouse.SetPosition(0, 0);
Console.WriteLine(Mouse.GetState().Position);`

I also tried with OpenTK.Input.Mouse, and same problem happens. It's a very big issue for FPS games where we want to reset the mouse position to the center of the game window at each loop.

Is there any alternatives for getting the mouse position on Mac?

**Comment by tomspilman** [comment]
@cra0zy @dellis1972

**Comment by cra0zy** [comment]
If you are using DesktopGL templates, #4755 indirectly fixes it.

**Comment by Torshid** [comment]
I found a temporary solution yesterday for FPS games (or 3D stuff needing the mouse to be at the center) which is avoiding to rely on the center. This solution works with either OpenTK (source of the problem) or MG:

```csharp
private bool SkipMouse = false;
private Vector2 PMPosition; //previous mouse position

protected override void Update(GameTime Time)
{
    MouseState MState = Mouse.GetState();
    Vector2 MPosition = new Vector2(MState.X, MState.Y) * 2; //need to multiply by 2 for mac retina only, loosing one bit of data btw, thanks opentk

    if (!SkipMouse)
    {
        float DiffX = (MPosition.X - this.PMPosition.X) / 2f;
        float DiffY = (MPosition.Y - this.PMPosition.Y) / 2f;

        //camera updates or whatever...
    }
    
    this.SkipMouse = false;

    //other updates

    //mouse position is put into the window bounds if it is outside
    //can be simplified but it's cool like this...
    int NewX = (int)(MPosition.X < 0 ? MPosition.X + this.Graphics.PreferredBackBufferWidth :
                MPosition.X > this.Graphics.PreferredBackBufferWidth ? MPosition.X - this.Graphics.PreferredBackBufferWidth : MPosition.X);
    int NewY = (int)(MPosition.Y < 0 ? MPosition.Y + this.Graphics.PreferredBackBufferHeight :
                MPosition.Y > this.Graphics.PreferredBackBufferHeight ? MPosition.Y - this.Graphics.PreferredBackBufferHeight : MPosition.Y);

    Mouse.SetPosition(NewX, NewY);

    if (NewX != MPosition.X || NewY != MPosition.Y)
    {
        this.SkipMouse = true; //skip the next mouse update
    }

    this.PMPosition = new Vector2(NewX, NewY);
}
```

(By the way this can be simplified a bit but I hope you understood the solution.)

**Off topic:**
I don't understand how such problems still occur after years of development, I mean if a class is to be free of any bugs, it surely would be the mouse... I was amazed by MG a few years ago and came back to it recently thanks to recent changes with Xamarin, but I am seriously being deceived, as there isn't even a proper UI supported (Xwt, Gtk, Eto, Winforms, ...). I am now considering the option of forgetting cross platform stuff for a while and reconcentrate on Windows.

To the MG team, do not take this as an offense as it is not, I really respect your work but there are important things to review in order to boost the power and importance of MG in the world of game development.

**Comment by tomspilman** [comment]
> By the way this can be simplified a bit but I hope you understood the solution

Yes...  but we're looking to take a different path in the next release.

Still the move to SDL2 will resolve the set mouse issues once and for all for the GL platforms.


> important things to review in order to boost the power and importance of MG in the world of game development.

We are fully aware of what is needed.

**Comment by cra0zy** [comment]
This can be closed, fixed with #4755
